src/service/cache-cid-api.js

```diff
@@ -0,0 +1,209 @@
+    this.apiKeyMap_ = null;
+
+    /** @private {?Promise<?string>} */
+    this.publisherCidPromise_ = null;
+
+    /** @private {!./timer-impl.Timer} */
+    this.timer_ = Services.timerFor(this.ampdoc_.win);
+  }
+
+  /**
+   * Returns true if the page is embedded in CCT and is served by a proxy.
+   * @returns {boolean}
+   */
+  isSupported() {
+    return this.isCctEmbedded() && this.viewer_.isProxyOrigin();
+  }
+
+  /**
+   * Returns scoped CID retrieved from the Viewer.
+   * @param {string} scope
+   * @return {!Promise<?string>}
+   */
+  getScopedCid(scope) {
+    // promise for whether we do stuff at all
+    if (!this.viewer_.isCctEmbedded()) {
+      return Promise.resolve(null);
+    }
+
+    const apiKey = this.isScopeOptedIn(scope);
+    if (!apiKey) {
+      return Promise.resolve(null);
+    }
+
+    if (!this.publisherCidPromise_) {
+      const url = CACHE_API_URL + SERVICE_KEY_;
+      this.publisherCidPromise_ = this.fetchCid_(url);
+    }
+
+    return this.publisherCidPromise_.then(publisherCid => {
+      return this.scopeCid_(publisherCid, scope);
+    });
+  }
+
+  /**
+   * Returns scoped CID retrieved from the Viewer.
+   * @param {string} url
+   * @return {!Promise<?string>}
+   */
+  fetchCid_(url) {
+    const payload = dict({
+      'publisherOrigin': getSourceOrigin(this.ampdoc_.win.location),
+    });
+
+    // Make the XHR request to the cache endpoint.
+    return this.timer_.timeoutPromise(
+        TIMEOUT,
+        Services.xhrFor(this.ampdoc_.win).fetchJson(url, {
+          method: 'POST',
+          ampCors: false,
+          credentials: 'include',
+          mode: 'cors',
+          body: payload,
+        }).then(res => {
+	  return res.json().then(response => {
+	    if (response['optOut']) {
+	      return null;
+	    }
+	    const cid = response['publisherClientId'];
+            if (!cid && response['alternateUrl']) {
+              // If an alternate url is provided, try again with the alternate url
+              // The client is still responsible for appending API keys to the URL.
+              const alt = `${response['alternateUrl']}?key=${SERVICE_KEY_}`;
+              return this.fetchCid_(dev().assertString(alt))
+                  .then(altRes => {
+                    return altRes.json().then(altResponse => {
+		      return altResponse['publisherClientId'] || null;
+		    });
+                  });
+	    }
+            return cid;
+	  });
+        }).catch(e => {
+          if (e && e.response) {
+            e.response.json().then(res => {
+              dev().error(TAG, JSON.stringify(res));
+            });
+          } else {
+            dev().error(TAG, e);
+          }
+          return null;
+        }));
+  }
+
+  /**
+   * Returns scoped CID extracted from the fetched publisherCid.
+   * @param {string} publisherCid
+   * @param {string} scope
+   * @return {?string}
+   */
+  scopeCid_(publisherCid, scope) {
+    if (!publisherCid) {
```

Comment:
`{string}` is non-nullable by default. Is it possible for this to be empty string or should we change the type?

Reply:
fetchCid_ returns a nullable string, which used to propagate to this method.  I moved the null check to the call site.